### PR TITLE
fix nested part body parsing, and new index-bodies related modifier options, other options doctype, timeout, static tags..

### PR DESCRIPTION
**Pull request by bitsofinfo**
- fix body indexing when nested parts are present
- `--index-bodies-ignore-content-types=[keyword,keyword2]` option
- `--index-bodies-html-parser=[html.parser | lxml | html5lib]`
- `--tags=key1=v1,key2=v2`
- `--es-document-type=custom`
- `--es-http-timeout-seconds=xxx`
- `--es-upload-attempts=N`

### CODE REVIEW

**Comment by oliver006** [comment]
Thanks for the PR! I'll try to get to it this weekend.

**Comment by bitsofinfo** [comment]
?

**Comment by oliver006** [comment]
Sorry, didn't get to that the last couple of days. I saw you pushed up a few more commits, is it RFR? I'll try to get to it soon.

**Comment by bitsofinfo** [comment]
RFR? not sure what that means. Its just more options/functionality

**Comment by oliver006** [comment]
RFR -> ready for review

**Comment by bitsofinfo** [comment]
yes

File: README.md [hunk 1]
```diff
@@ -328,6 +328,12 @@ GET _search
         }
 ```
 
+#### Resources
+
+Not interested in figuring out how to setup Elasticsearch and Kibana? Take a look at [comms-analyzer-toolbox](https://github.com/bitsofinfo/comms-analyzer-toolbox) which bundles
```
Comment (oliver006) [inline README.md]: This section doesn't really belong here, can you please remove it?

File: src/index_emails.py [hunk 2]
```diff
@@ -82,24 +84,74 @@ def create_index():
...
         upload_data_txt += json.dumps(cmd) + "\n"
         upload_data_txt += json.dumps(item) + "\n"
 
-    request = HTTPRequest(tornado.options.options.es_url + "/_bulk", method="POST", body=upload_data_txt, request_timeout=240)
-    response = http_client.fetch(request)
-    result = json.loads(response.body)
+    upload_successful = False
+    upload_attempts = 0
+    res_txt = ""
+    while ((not upload_successful) and (upload_attempts < tornado.options.options.es_upload_attempts)):
```
Comment (oliver006) [inline src/index_emails.py:94]: Why is this retry logic needed?

Comment (oliver006) [reply src/index_emails.py:94]: generally I prefer something like
```
attempts_left = 10
while attempts_left:
    # do something
    if success:
         break
    attempts_left -= 1
```

File: src/index_emails.py [hunk 3]
```diff
@@ -213,7 +269,25 @@ def load_from_file():
                            help="Number of shards for ES index")
 
     tornado.options.define("index_bodies", type=bool, default=False,
-                           help="Will index all body content, stripped of HTML/CSS/JS etc. Adds fields: 'body' and 'body_size'")
+                           help="Will index all body content, stripped of HTML/CSS/JS etc. Adds fields: 'body', 'body_size' and 'body_filenames' for any multi-part attachments")
+
+    tornado.options.define("index_bodies_ignore_content_types", default="application,image",
```
Comment (oliver006) [inline src/index_emails.py:276]: `index_bodies_ignore_content_types` is very long - think there's a shorter name that describes what's going on?

File: src/index_emails.py [hunk 4]
```diff
@@ -82,24 +84,74 @@ def create_index():
...
+
 def normalize_email(email_in):
     parsed = email.utils.parseaddr(email_in)
     return parsed[1]
 
+def content_type_is_ignored(content_type):
+    if tornado.options.options.index_bodies_ignore_content_types:
+        ignore_types = tornado.options.options.index_bodies_ignore_content_types.split(",")
+        for to_ignore in ignore_types:
+            if to_ignore:
```
Comment (oliver006) [inline src/index_emails.py]: combine this with the `if` in the next line?

File: src/index_emails.py [hunk 5]
```diff
@@ -82,24 +84,74 @@ def create_index():
...
+            if mpart is not None: # recurse
+                process_msg_body(mpart,result)
+
+    else:
+        # log filename of any part dispostiion
+        if msg.get_filename():
+            result['body_filenames'] += (msg.get_filename() + " ")
+
+        # ignore the content-type?
+        if content_type_is_ignored(msg.get_content_type()):
```
Comment (oliver006) [inline src/index_emails.py:146]: this is odd.
It says `if content_type_is_ignored` and then you take action, almost like it's the wrong way around.
Is this more like `content_type_included` ?
Can you name the function more clearly?

File: src/index_emails.py [hunk 6]
```diff
@@ -82,24 +84,74 @@ def create_index():
...
+            result['body_ignored_content_types'] += (msg.get_content_type() + " ")
```
Comment (oliver006) [inline src/index_emails.py:147]: `(msg.get_content_type() + " ")` - don't think the enclosing parenthesis are needed?

File: src/index_emails.py [hunk 7]
```diff
@@ -82,24 +84,74 @@ def create_index():
...
+def process_msg_body(msg,result):
+    if msg.is_multipart():
+        for mpart in msg.get_payload(): # don't decode
+            if mpart is not None: # recurse
+                process_msg_body(mpart,result)
+
+    else:
+        # log filename of any part dispostiion
+        if msg.get_filename():
+            result['body_filenames'] += (msg.get_filename() + " ")
```
Comment (oliver006) [inline src/index_emails.py:143]: don't think the enclosing parenthesis are needed

**Review by oliver006 — COMMENTED** [review]
Thanks for the PR (and sorry it took me a while to get to this) - I left you a couple of review notes, let me know if you have questions.

File: src/index_emails.py (hunk 3 above)
Comment (bitsofinfo) [reply src/index_emails.py:276]: Will consider it in next iteration, working on other features.

File: src/index_emails.py (hunk 6 above)
Comment (bitsofinfo) [reply src/index_emails.py:147]: Will consider it in next iteration, working on other features.

File: src/index_emails.py (hunk 5 above)
Comment (bitsofinfo) [reply src/index_emails.py:146]: Function is named appropriately, if the content type is in the ignored list, we add it to a field called `body_ignored_content_types` to show that the body was excluded due to the content type matching that argument

File: src/index_emails.py (hunk 7 above)
Comment (bitsofinfo) [reply src/index_emails.py:143]: Will consider it in next iteration, working on other features.

File: src/index_emails.py (hunk 4 above)
Comment (bitsofinfo) [reply src/index_emails.py]: Will consider it in next iteration, working on other features.

File: src/index_emails.py (hunk 2 above)
Comment (bitsofinfo) [reply src/index_emails.py:94]: Permits flexibility for transient network failures or load in ES where the request fails to complete within timeout.  Such as `Cannot assign requested address` failures, something with the http client lib you are using is producing for very large mbox files that take a while to run. Looking more into it.

File: README.md (hunk 1 above)
Comment (bitsofinfo) [reply README.md]: Relevant, just a simple link to a tool that makes it easier to use the functionality of this project.

File: src/index_emails.py (hunk 2 above)
Comment (oliver006) [reply src/index_emails.py:94]: Ok

File: README.md (hunk 1 above)
Comment (oliver006) [reply README.md]: Sure but I'm not affiliated with that project so I prefer to remove the link.

**Comment by oliver006** [comment]
Thanks for your contributions!

I left a few comments about how to improve some of the changes.

In general, I' prefer to wrap up this PR and not turn it into an open-ended stream of changes. If you have future features coming that's great but they can be handled in a separate PR.

**Comment by bitsofinfo** [comment]
this should al be ready to merge now, anything else Ill put in other PRs

File: src/index_emails.py (hunk 2 above)
Comment (oliver006) [reply src/index_emails.py:94]: If you don't intend to fix/improve this section in this PR, can you please revert the changes in `upload_batch()`?

File: src/index_emails.py [hunk 8]
```diff
@@ -82,24 +84,74 @@ def create_index():
...
 def normalize_email(email_in):
     parsed = email.utils.parseaddr(email_in)
     return parsed[1]
 
+def content_type_is_ignored(content_type):
+    if tornado.options.options.index_bodies_ignore_content_types:
+        ignore_types = tornado.options.options.index_bodies_ignore_content_types.split(",")
+        for to_ignore in ignore_types:
+            if to_ignore:
+                if to_ignore.strip().lower() in content_type.lower():
```
Comment (oliver006) [inline src/index_emails.py]: Should move the `lower()` out of the loop so it's not called for each call to `content_type_is_ignored()`.

File: src/index_emails.py (hunk 5 above)
Comment (oliver006) [reply src/index_emails.py:146]: 👍

File: src/index_emails.py [hunk 9]
```diff
@@ -135,17 +187,21 @@ def convert_msg_to_json(msg):
...
+        process_msg_body(msg,result)
 
+        # log body size
         result['body_size'] = len(result['body'])
 
+    # tags?
+    if tornado.options.options.tags:
+        kv_pairs = tornado.options.options.tags.split(",")
+        for kv_pair in kv_pairs:
+            kv = kv_pair.split("=")
```
Comment (oliver006) [inline src/index_emails.py:203]: `kv` needs a `len(kv) == 2` check before it can be used.

**Review by oliver006 — CHANGES_REQUESTED** [review]
Thanks for the changes. Please see my comments for the addtl changes needed before we can merge this.

**Comment by bitsofinfo** [comment]
Lost, improve what in upload batch? it works fine

File: src/index_emails.py (hunk 8 above)
Comment (bitsofinfo) [reply src/index_emails.py]: done

File: src/index_emails.py (hunk 9 above)
Comment (bitsofinfo) [reply src/index_emails.py:203]: done

File: src/index_emails.py [hunk 10]
```diff
@@ -123,10 +123,11 @@ def normalize_email(email_in):
 
 def content_type_is_ignored(content_type):
     if tornado.options.options.index_bodies_ignore_content_types:
+        content_type_lcase = content_type.lower()
         ignore_types = tornado.options.options.index_bodies_ignore_content_types.split(",")
         for to_ignore in ignore_types:
             if to_ignore:
-                if to_ignore.strip().lower() in content_type.lower():
+                if to_ignore.strip().lower() in content_type_lcase:
```
Comment (oliver006) [inline src/index_emails.py]: Using `content_type_lcase` is a good start but what I really meant is `to_ignore.strip().lower()` should be done at startup time of the program, not on every invocation of this function.

File: src/index_emails.py [hunk 11]
```diff
@@ -200,7 +201,8 @@ def convert_msg_to_json(msg):
         kv_pairs = tornado.options.options.tags.split(",")
         for kv_pair in kv_pairs:
             kv = kv_pair.split("=")
-            result[kv[0]] = kv[1]
+            if len(kv) == 2:
+                result[kv[0]] = kv[1]
```
Comment (oliver006) [inline src/index_emails.py]: 👍

**Comment by oliver006** [comment]
To clarify my comment regarding `upload_batch()`: if you don't want to clean up the logic as per my comment and/or remove the retry logic I prefer that you undo your changes to that function.

**Comment by bitsofinfo** [comment]
Will just use my fork, thanks

**Comment by oliver006** [comment]
Sorry to hear that.

**Pull request closed without merging**